src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineWindowsLauncher.java

```diff
@@ -0,0 +1,269 @@
+                    return; // failed to connect
+                }
+            } else {
+                logWarning(computer, listener, "bootstrapresult failed");
+                return;
+            }
+            conn = cleanupConn;
+
+            SCPClient scp = conn.createSCPClient();
+            String tmpDir = "C:\\";
+            logInfo(computer, listener, "Copying slave.jar to: " + tmpDir);
+            scp.put(Jenkins.getInstance().getJnlpJars("slave.jar").readFully(), "slave.jar", tmpDir);
+
+            // Confirm Java is installed
+            if (!testCommand(computer, conn, "java -fullversion", logger, listener)) {
+                logWarning(computer, listener, "Java is not installed.");
+            }
+
+            //TODO: allow jvmopt configuration
+            String launchString = "java -jar C:\\slave.jar";
+
+            logInfo(computer, listener, "Launching Jenkins agent via plugin SSH: " + launchString);
+            final Session sess = conn.openSession();
+            sess.execCommand(launchString);
+            computer.setChannel(sess.getStdout(), sess.getStdin(), logger, new Channel.Listener() {
+                @Override
+                public void onClosed(Channel channel, IOException cause) {
+                    sess.close();
+                    conn.close();
+                }
+            });
+        } catch (Exception e) {
+            logException(computer, listener, "Error getting exception", e);
+        }
+    }
+
+    private boolean testCommand(ComputeEngineComputer computer, Connection conn, String checkCommand, PrintStream logger, TaskListener listener)
+            throws IOException, InterruptedException {
+        logInfo(computer, listener, "Verifying: " + checkCommand);
+        return conn.exec(checkCommand, logger) == 0;
+
+    }
+
+    private boolean authenticateSSH(WindowsConfiguration windowsConfig, Connection sshConnection,
+                                    TaskListener listener) throws Exception{
+        boolean isAuthenticated;
+        String windowsUsername = windowsConfig.getWindowsUsername();
+        if (windowsConfig.getPrivateKeyCredentialsId() != null) {
+            StandardUsernameCredentials cred = windowsConfig.getPrivateKeyCredentials();
+            isAuthenticated = SSHAuthenticator.newInstance(sshConnection, cred, windowsUsername).authenticate(listener);
+        } else {
+            isAuthenticated = sshConnection.authenticateWithPassword(windowsUsername, windowsConfig.getPassword());
+        }
+        return isAuthenticated;
+    }
+
+    private boolean bootstrap(ComputeEngineComputer computer, TaskListener listener) throws IOException,
+            Exception { //TODO: better exceptions
+        logInfo(computer, listener, "bootstrap");
+        ComputeEngineInstance node = computer.getNode();
+        WindowsConfiguration windowsConfig = node.windowsConfig;
+        if (node == null) {
```

Comment:
This can be done more concisely with: Preconditions.checkNotNull()
see: https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Preconditions.html#checkNotNull-T-

Reply:
tracking issue/part of refactoring